host-local: Update host-local IPAM to support Windows

rakelkar opened the pull request:
Get host-local to work for Windows. This is required for Flannel VXLAN mode on Windows.

squeed commented on plugins/ipam/host-local/backend/disk/backend.go:
```diff
@@ -56,6 +57,10 @@ func New(network, dataDir string) (*Store, error) {
 func (s *Store) Reserve(id string, ip net.IP, rangeID string) (bool, error) {
 	fname := filepath.Join(s.dataDir, ip.String())
+	if runtime.GOOS == "windows" {
+		fname = strings.Replace(fname, ":", "_", -1)
```
Hah, good catch!

squeed commented on pkg/ip/route_unspecified.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-// +build !linux
+// +build !linux,!windows
```
Is there supposed to be a Windows version of this file? Or am I just missing something?

rakelkar replied:
Cant make a windows specific version of these methods since they reference the netlink package in their method signature "dev netlink.Link" ... Not sure why this file is present in the repo.

squeed commented on pkg/ipam/ipam_windows.go:
```diff
@@ -0,0 +1,38 @@
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+func ExecAdd(plugin string, netconf []byte) (types.Result, error) {
```
It looks like these functions are duplicated - Can you just move ConfigureIface to a build-specific file?

rakelkar replied:
Moved the execXX funcs to exec.go - let me know if thats not what you had in mind

squeed commented on plugins/ipam/host-local/backend/disk/lock.go:
```diff
@@ -1,3 +1,5 @@
+// +build linux
```
Just do !windows here - this "should" work on all Unix-like systems.

rakelkar replied:
Done

squeed commented on pkg/ip/addr.go:
```diff
@@ -1,3 +1,5 @@
+// +build linux
```
Out of curiosity, does Windows also need to worry about "tentative" addresses?

rakelkar replied:
Not sure, asking around

rakelkar replied:
So it seems Windows doesn't need to worry about it right now since HNS doesn't support IPv6 yet - so ja bitte for now I guess ;-)

squeed commented:
Seems pretty reasonable to me. I added a first pass of changes - you're right, not too many changes needed.

When doing build-tagged files, it's considered good practice to only duplicate the minimum amount of os-dependent code, and put the rest in non-build-tagged files.

rakelkar commented:
@squeed Have addressed your comments, could you take a look? The test failure on CI is unrelated to this change (possibly a flaky test?).

squeed commented:
Kicked that CI job and it's green now, dammit.

Seems reasonable to me. I want to think for one more day if we want to have a specific conditional build policy - the maintainer meeting is tomorrow.

rosenhouse commented on pkg/ipam/ipam_windows.go:
```diff
@@ -0,0 +1,28 @@
+// +build windows
```
I don't believe this line is necessary, since the filename ends in `_windows.go`.

[[Reference](https://golang.org/pkg/go/build/#hdr-Build_Constraints)]

rakelkar replied:
removed

rosenhouse commented on plugins/ipam/host-local/backend/disk/lock_windows.go:
```diff
@@ -0,0 +1,61 @@
+		lockPath = path.Join(lockPath, "lock")
+	}
+
+	f, err := filemutex.New(lockPath)
```
this code seems identical to the POSIX version, except it makes use of this nice cross-platform adapter.  Could we make this file the one true `lock.go` and remove the POSIX-only implementation entirely?

rakelkar replied:
Done

rosenhouse commented on plugins/ipam/host-local/backend/disk/lock_windows.go:
```diff
@@ -0,0 +1,61 @@
+package disk
+
+import (
+	"github.com/alexflint/go-filemutex"
```
Is this package added to the `vendor` directory somewhere?  I don't see it.

rosenhouse replied:
[meta comment: I'm surprised this passed Travis without doing that.    Maybe we should open a separate PR to enforce vendoring at CI time.]

rakelkar replied:
Vendored

rakelkar commented:
Travis doesn't build windows... I can add a build.ps1 and an appveyor yml
to kick it off? The ps1 would only include plugins that work for Windows -
so only host-local for now...

On Oct 17, 2017 6:39 AM, "Gabe Rosenhouse" <notifications@github.com> wrote:

> *@rosenhouse* commented on this pull request.
> ------------------------------
>
> In plugins/ipam/host-local/backend/disk/lock_windows.go
> <https://github.com/containernetworking/plugins/pull/77#discussion_r145132704>
> :
>
> > +// Licensed under the Apache License, Version 2.0 (the "License");
> +// you may not use this file except in compliance with the License.
> +// You may obtain a copy of the License at
> +//
> +//     http://www.apache.org/licenses/LICENSE-2.0
> +//
> +// Unless required by applicable law or agreed to in writing, software
> +// distributed under the License is distributed on an "AS IS" BASIS,
> +// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> +// See the License for the specific language governing permissions and
> +// limitations under the License.
> +
> +package disk
> +
> +import (
> +	"github.com/alexflint/go-filemutex"
>
> [meta comment: I'm surprised this passed Travis without doing that. Maybe
> we should open a separate PR to enforce vendoring at CI time.]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/containernetworking/plugins/pull/77#discussion_r145132704>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADQphave4hcpc8WYnFzV4hBgHG5xqcokks5stK4sgaJpZM4PxM90>
> .
>

rosenhouse commented on plugins/ipam/host-local/backend/disk/lock.go:
```diff
@@ -15,36 +15,47 @@
-// Close closes underlying file
 func (l *FileLock) Close() error {
-	return l.f.Close()
+	return nil
```
why not close the file?

rosenhouse replied:
oh, because the library doesn't expose a `Close()` method at all.  uh oh....

rosenhouse replied:
Although apparently we're not using `Close()` anyhow.  When I delete the `Close()` method from the `backend.Store` interface, the program still compiles.  So I guess you could just do that.

rakelkar replied:
Nice catch - will remove, though I must say with a touch of disquiet.

rakelkar replied:
So it seems like main.go:51 creates a Store struct and not the interface and calls defer store.Close so close is called after all...

rakelkar replied:
There is an open PR on the library to add a close method, will create a PR to update this plugin once that has been merged. alexflint/go-filemutex#2

rakelkar replied:
Updated with shiny new close method!

rosenhouse commented on plugins/ipam/host-local/backend/disk/lock.go:
```diff
@@ -15,36 +15,47 @@
 func (l *FileLock) Lock() error {
-	return syscall.Flock(int(l.f.Fd()), syscall.LOCK_EX)
+	l.f.Lock()
+	return nil
```
I didn't realize before that this library `panic`s on errors rather than returning them. 
Effective Go says that ["real library functions should avoid panic"](https://golang.org/doc/effective_go.html#panic).  I wonder if that library deserves a PR...

rakelkar commented on plugins/ipam/host-local/host_local_test.go:
```diff
@@ -37,34 +38,34 @@ var _ = Describe("host-local Operations", func() {
+					{"dst": "2001:db8:2::0/64", "gw": "2001:db8:3::1"}
+				]
+			}
+		}`, tmpDir, tmpDir)
```
I ran gofmt - and it felt indents were in order!

rakelkar commented on plugins/ipam/host-local/host_local_test.go:
```diff
@@ -517,6 +518,15 @@ var _ = Describe("host-local Operations", func() {
+func getTmpDir() (string, error) {
+	tmpDir, err := ioutil.TempDir("", "host_local_artifacts")
+	if err == nil {
+		tmpDir = filepath.ToSlash(tmpDir)
```
so without this the un-escaped backslashes in the windows path cause JSON parsing to barf

rakelkar commented on plugins/ipam/host-local/backend/disk/backend.go:
```diff
@@ -56,10 +56,7 @@ func New(network, dataDir string) (*Store, error) {
-	if runtime.GOOS == "windows" {
-		fname = strings.Replace(fname, ":", "_", -1)
-	}
+	fname := GetEscapedPath(s.dataDir, ip.String())
```
Turns out I missed a bunch, also escaping the colon in the dataDir doesnt work so well when the path includes a drive letter - duh!

rakelkar commented:
@rosenhouse Have addressed all comments, got the shiny new changes from alexflint and made fixes to get tests to run on Windows. Let me know if you see something anything don't like. The current build failure is an unrelated flaky test.

rosenhouse commented:
@rakelkar thanks for all your hard work on this!  we're excited to merge this, just as soon as we have Continuous integration set up for Windows in this repo.  I've got PR #84 open, but its not yet green.  Once that's merged, then we can return to this PR.

rosenhouse commented:
@rakelkar now that #84 is merged, please rebase and squash this down to a single commit, e.g. something like https://github.com/rosenhouse/plugins/commit/799776fec6216a76310df96ec486b7ca3dd44cbf

rakelkar commented:
@rosenhouse rebased and squashed!

rosenhouse commented on plugins/ipam/host-local/backend/disk/lock.go:
```diff
@@ -16,36 +15,45 @@
+
+	if fi.IsDir() {
+		lockPath = path.Join(lockPath, "lock")
+	}
```
is this line necessary for windows?  If so, maybe it deserves a unit test?

rakelkar replied:
yes it is required - added unit tests for filepath and folderpath

rakelkar replied:
LMK if you like what you see, will wait for LGTM to squash!

rakelkar replied:
squashed commits

rakelkar commented:
squashed commits
